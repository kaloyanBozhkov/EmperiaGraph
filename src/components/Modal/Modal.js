import React, { useEffect } from 'react'
import { connect } from 'react-redux'

import styles from './styles.module.scss'
import Icon from 'UI/Icon/Icon'

import ConfirmPassword from './ConfirmPassword/ConfirmPassword'

import { closeModal } from '~/redux/modal/modal.actions'

/**
 * @param  {string} {openedModal -> name of modal to open
 * @param  {obj} data -> data payload, can have anything including onBtnClick
 * @param  {fn} onCloseModal -> fn to call in order to close modal
 */
export const Modal = ({ modal, data = {}, onCloseModal }) => {

  // handle escape for modal close
  useEffect(() => {

    // on mount bind listener
    const listener = document.addEventListener('keydown', (event) => {
      if (event.key === 'Escape') {
        onCloseModal()
      }
    })

    // on unmount unbind listener
    return () => document.removeEventListener('keydown', listener)
  })

  // if modal is null dont show anything
  if (!modal) {
    return null
  }

  const { modalLabel, modalModifier, ...payload } = data

  let content = null

  switch (modal) {
    case 'confirmPassword': {
      const { onSave, ...otherProps } = payload
      const onSaveWithClose = (...args) => {
        // after getting password from input and passing it as arg, close modal
        onCloseModal()
        // run the onSave from the modal button click, passing all arguments
        onSave(...args)
      }

      content = <ConfirmPassword onSave={onSaveWithClose} {...otherProps} />
      break
    }
    default:
      content = <p>Modal is empty, not configured!</p>
      break
  }

  return (
    <>
      <div className={styles.modalOverlay} />
      <div className={[styles.box, (modalModifier && styles[modalModifier]) || undefined].join(' ').trim()}>
        {modalLabel && <p>{modalLabel}</p>}
        <div className={styles.close} onClick={onCloseModal}>
          <Icon icon="times" />
        </div>
        {content}
      </div>
    </>
  )
}

const mapStateToProps = (state) => ({
  modal: state.modalReducer.modal,
  data: state.modalReducer.data,
  error: state.modalReducer.error,
})

const mapDispatchToProps = (dispatch) => ({
  onCloseModal: () => dispatch(closeModal()),
})

export default connect(mapStateToProps, mapDispatchToProps)(Modal)
