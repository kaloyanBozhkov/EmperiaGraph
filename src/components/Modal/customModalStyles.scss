@import 'mixins';

// modal modifier - for the modal heading
@mixin customModalModifiers() {
}
