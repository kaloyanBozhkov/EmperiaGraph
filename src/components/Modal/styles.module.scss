@import 'variables';
@import 'mixins';
@import 'keyframes';
@import './customModalStyles.scss';

.modalOverlay {
  @include animate(opacityIn, 300ms);
  position: fixed;
  background-color: rgba(0, 0, 0, 0.7);
  z-index: 50;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
}

.box {
  @include animate(popIn, 300ms);
  position: fixed;
  z-index: 55;
  background-color: white;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  min-width: 40rem;
  padding: 2rem;

  // share close icon
  .close {
    position: absolute;
    top: 1.5rem;
    right: 2rem;
    color: $primary-color;
    font-weight: bolder;
    font-size: 1.4rem;
    cursor: pointer;
  }

  .title {
    padding-left: .5rem;
    border-left: 3px solid $primary-color;
    font-size: 1.4rem;
  }

  @include customModalModifiers();
}