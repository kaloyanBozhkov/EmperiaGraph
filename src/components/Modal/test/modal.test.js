import React from 'react'
import { shallow, mount } from 'enzyme'
import { Modal } from '../Modal'

describe('testing modal', () => {
  it('should render modal', () => {
    const modal = shallow(<Modal modal="someUnhandledModal" />)
    expect(modal).toMatchSnapshot()
  })
  it('should fire close fn on cose icon click', () => {
    const fn = jest.fn(() => 1)
    const modal = shallow(<Modal modal="someUnhandledModal" onCloseModal={fn} />)
    modal.find('.close').at(0).simulate('click')
    expect(fn).toHaveBeenCalled()
  })
  it('should render modal with defalt "modal is empty" msg if modal name provided is not handled', () => {
    const modal = shallow(<Modal modal="unhandledModalName" />)

    expect(modal.find('.box p').at(0).html()).toEqual('<p>Modal is empty, not configured!</p>')
  })
  it('should return null if modal not specified', () => {
    const modal2 = shallow(<Modal modal={null} />)
    expect(modal2).toEqual({})
  })
  it('should render cofirm password modal with expected contents and pass them the expected props', () => {
    const props = {
      text: 'hello world',
    }
    const modal = shallow(<Modal data={props} modal="confirmPassword" />)
    expect(modal).toMatchSnapshot()
    expect(modal.find('ConfirmPassword').prop('text')).toEqual('hello world')
  })

  it('should close modal when pressed esc', () => {
    const closeFn = jest.fn()

    // obj to keep track of eventlisteners
    const map = {}

    // replace addEventListener with jest fn
    document.addEventListener = jest.fn((event, cb) => {
      
      // keep ref to event
      map[event] = cb
    })

    const modal = mount(<Modal modal="someModalName" onCloseModal={closeFn} />)

    // simulate keypress
    map.keydown({ key: 'Escape'})

    expect(closeFn).toHaveBeenCalled()
  })
})
