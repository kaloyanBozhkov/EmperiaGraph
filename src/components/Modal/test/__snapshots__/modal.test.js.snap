// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing modal should render cofirm password modal with expected contents and pass them the expected props 1`] = `
<Fragment>
  <div
    className="modalOverlay"
  />
  <div
    className="box"
  >
    <div
      className="close"
    >
      <Icon
        icon="times"
      />
    </div>
    <ConfirmPassword
      onSave={[Function]}
      text="hello world"
    />
  </div>
</Fragment>
`;

exports[`testing modal should render modal 1`] = `
<Fragment>
  <div
    className="modalOverlay"
  />
  <div
    className="box"
  >
    <div
      className="close"
    >
      <Icon
        icon="times"
      />
    </div>
    <p>
      Modal is empty, not configured!
    </p>
  </div>
</Fragment>
`;
